lib/fluent/plugin/in_tail.rb

```diff
@@ -669,6 +710,51 @@ def on_timer
       end
     end
 
+    class TailThread
+      class Pool
+        def initialize(max_size, &session)
+          @max_size = max_size
+          @queue = Queue.new
+          @threads = []
+          @running = false
+          session.call(self)
+          @running = true
+        ensure
+          stop
+        end
+
+        def run(&task)
+          @queue.push(task)
+          @threads << create_thread if @threads.size < @max_size
+        end
+
+        def stop
+          return unless @running
+          @running = false
+
+          until @queue.num_waiting == @threads.size
+            sleep 0.01
+          end
+          begin
+            Timeout.timeout(1) do
+              @threads.each{|th| th.join }
```

Comment:
I'd tried to more proper queue termination in: https://github.com/fluent/fluentd/pull/3185/commits/382df6cfcdba3341b95aa5efd3050e0a398199ca.
But with this patch, I got still `terminating event_loop forcedly` errors.

Reply:
On testing, in_tail plugin instance is always blocked for termination by `shutdown: true` (default driver#run keyword arguments).
This is normal behavior on `in_tail` during tailing and emitting events phase.
`in_tail` does not support immediate stop on tailing files left.
 
In https://github.com/fluent/fluentd/pull/3185/commits/2d67e44608408cea32f7abfdfe2f2447e3915243, I specified `shutdown: false` to prevent hard timeout messages generating.